### PR TITLE
Ifpack2: add AVG Combine Mode option to Schwarz DD

### PR DESCRIPTION
**Pull request by rstumin**
Issue:  #10901
Tests: Added to MueLu
Stakeholder feedback: n/a

Auto-PR for SHA be3d516

### CODE REVIEW

File: packages/muelu/test/scaling/CMakeLists.txt [hunk 1]
```diff
@@ -289,6 +289,14 @@ IF (${PACKAGE_NAME}_HAVE_TPETRA_SOLVER_STACK)
     COMM mpi # HAVE_MPI required
   )
 
+  TRIBITS_ADD_TEST(
+    Driver
+    NAME "SchwarzSmoothingWithAverages"
+    ARGS "--linAlgebra=Tpetra --xml=oLapSchwarzEquivBlkSmoo.xml --belosType=\"Fixed\ Point\" --rowmap=GblkMap.dat --matrix=GblkAmat.dat --rhs=GblkRhs.dat --tol=.1"
```
Comment (GrahamBenHarper) [inline packages/muelu/test/scaling/CMakeLists.txt:295]: Does this need the `regularOverlap.dat` file anywhere? I see it referenced in comments, and I see it's copied to the build dir, but I don't see it explicitly called by anything.

**Review by csiefer2 — DISMISSED** [review]

**Review by GrahamBenHarper — APPROVED** [review]
I'm surprised this didn't exist before. It looks useful!

File: packages/muelu/test/scaling/CMakeLists.txt (hunk 1 above)
Comment (rstumin) [reply packages/muelu/test/scaling/CMakeLists.txt:303]: first I thought I added the wrong test to CMakefile so then I pushed the "other test" ... now I realize that I actually had the test I wanted. Anyway, I just pushed a second change to the PR that now does both tests: averaging with block relaxation and averaging with schwarz. the block relaxation one uses the regularOverlap.dat file to do user defined blocks that just happen to correspond to the overlapping schwarz blocks.  Thanks for catching this potentially confusing issue.

Comment (csiefer2) [reply packages/muelu/test/scaling/CMakeLists.txt:303]: See @rstumin 's push

**Review by csiefer2 — DISMISSED** [review]

**Review by GrahamBenHarper — DISMISSED** [review]

File: packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp [hunk 2]
```diff
@@ -425,6 +425,19 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
...
+    // the number of procs who have a copy of each overlap dof
+    Teuchos::ArrayRCP<double>  dataNumOverlapCopies;
+    if (IsOverlapping_ && AvgOverlap_) {
+      if (num_overlap_copies_.get()  == nullptr) {
+        num_overlap_copies_.reset (new MV (Y.getMap (), 1, false));
+        RCP<MV> onesVec( new MV(OverlappingMatrix_->getRowMap(), 1, false) );
+        onesVec->putScalar(Teuchos::ScalarTraits<scalar_type>::one());
+        rcp_dynamic_cast<OverlappingRowMatrix<row_matrix_type>> (OverlappingMatrix_)->exportMultiVector (*onesVec, *(num_overlap_copies_.get ()), CombineMode_);
+      }
+      dataNumOverlapCopies = num_overlap_copies_.get ()->getDataNonConst(0);
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:438]: It's weird that only the Zoltan2Sphynx [build](https://trilinos-cdash.sandia.gov/viewBuildError.php?buildid=396529) fails for SC=complex<double>.

File: packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp [hunk 3]
```diff
@@ -550,6 +563,14 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
            << "OverlappingMatrix_ is null when it shouldn't be.  "
            "Please report this bug to the Ifpack2 developers.");
         OverlappingMatrix_->exportMultiVector (*OverlappingY, *C, CombineMode_);
+
+        // average solution in overlap regions if requested via "schwarz: combine mode" "AVG"
+        if (AvgOverlap_) {
+          Teuchos::ArrayRCP<double>  dataC = C->getDataNonConst(0);
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:569]: It's weird that only the Zoltan2Sphynx [build](https://trilinos-cdash.sandia.gov/viewBuildError.php?buildid=396529) fails for SC=complex<double>.

**Review by jhux2 — COMMENTED** [review]
See my comments.

**Review by mayrmt — COMMENTED** [review]
From a testing perspective, a change to `Ifpack2` should be tested with a unit test in `Ifpack2`.

**Review by jhux2 — COMMENTED** [review]
@rstumin I agree that it's better if the unit test is in Ifpack2.

File: packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp (hunk 2 above)
Comment (jhux2) [reply packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:438]: This is being tracked in https://github.com/trilinos/Trilinos/issues/10823#issuecomment-1206856350.

File: packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp (hunk 3 above)
Comment (jhux2) [reply packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:569]: This is being tracked in https://github.com/trilinos/Trilinos/issues/10823#issuecomment-1206856350.

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**